File: sdk/go/vault.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
)

type Secret struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Scope     string `json:"scope"`
	Last4     string `json:"last4"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

type CreateSecretInput struct {
	Name  string `json:"name"`
	Scope string `json:"scope"`
	Value string `json:"value"`
}

type UpdateSecretInput struct {
	Scope *string `json:"scope,omitempty"`
	Value *string `json:"value,omitempty"`
}

type VaultGetResult struct {
	Value string `json:"value"`
}

type Vault struct{ http *httpClient }

// Get fetches the value of a vault secret by name (agent-scoped endpoint).
func (r *Vault) Get(ctx context.Context, name string) (*VaultGetResult, error) {
	var out VaultGetResult
	err := r.http.do(ctx, http.MethodGet, "/v1/vault/"+url.PathEscape(name), nil, nil, &out, true)
	return &out, err
}

func (r *Vault) ListSecrets(ctx context.Context, tenantID string) ([]Secret, error) {
	var env struct {
		Items []Secret `json:"items"`
		Data  []Secret `json:"data"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/tenants/"+url.PathEscape(tenantID)+"/secrets", nil, nil, &env, true); err != nil {
		return nil, err
	}
	if env.Items != nil {
		return env.Items, nil
	}
	return env.Data, nil
}

func (r *Vault) CreateSecret(ctx context.Context, tenantID string, in CreateSecretInput) (*Secret, error) {
	var out Secret
	err := r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/secrets", nil, in, &out, false)
	return &out, err
}

func (r *Vault) UpdateSecret(ctx context.Context, tenantID, id string, in UpdateSecretInput) (*Secret, error) {
	var out Secret
	err := r.http.do(ctx, http.MethodPatch, "/v1/tenants/"+url.PathEscape(tenantID)+"/secrets/"+url.PathEscape(id), nil, in, &out, false)
	return &out, err
}

func (r *Vault) RevealSecret(ctx context.Context, tenantID, id string) (*VaultGetResult, error) {
	var out VaultGetResult
	err := r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/secrets/"+url.PathEscape(id)+"/reveal", nil, struct{}{}, &out, false)
	return &out, err
}

func (r *Vault) DeleteSecret(ctx context.Context, tenantID, id string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/tenants/"+url.PathEscape(tenantID)+"/secrets/"+url.PathEscape(id), nil, nil, nil, true)
}

File: sdk/js/node/src/vault.ts
import type { HttpClient } from "./client.js";

export interface Secret {
  id: string;
  name: string;
  scope: string;
  last4: string;
  created_at: string;
  updated_at: string;
}

export interface CreateSecretInput {
  name: string;
  scope: string;
  value: string;
}

export interface UpdateSecretInput {
  scope?: string;
  value?: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
}

export class Vault {
  constructor(private readonly http: HttpClient) {}

  /** Agent-scoped: fetch the value of a vault secret by name. */
  get(name: string): Promise<{ value: string }> {
    return this.http.get<{ value: string }>(`/v1/vault/${encodeURIComponent(name)}`);
  }

  /** Admin: list secrets for a tenant (values are masked to last 4 chars). */
  async listSecrets(tenantId: string): Promise<Secret[]> {
    const res = await this.http.get<ListEnvelope<Secret>>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/secrets`,
    );
    return res.items ?? res.data ?? [];
  }

  /** Admin: create a new secret for a tenant. */
  createSecret(tenantId: string, input: CreateSecretInput): Promise<Secret> {
    return this.http.post<Secret>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/secrets`,
      input,
    );
  }

  /** Admin: update scope or rotate the value of a secret. */
  updateSecret(tenantId: string, id: string, input: UpdateSecretInput): Promise<Secret> {
    return this.http.patch<Secret>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/secrets/${encodeURIComponent(id)}`,
      input,
    );
  }

  /** Admin: reveal the full plaintext value of a secret. */
  revealSecret(tenantId: string, id: string): Promise<{ value: string }> {
    return this.http.post<{ value: string }>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/secrets/${encodeURIComponent(id)}/reveal`,
      {},
    );
  }

  /** Admin: delete a secret. */
  deleteSecret(tenantId: string, id: string): Promise<void> {
    return this.http.delete<void>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/secrets/${encodeURIComponent(id)}`,
    );
  }
}
